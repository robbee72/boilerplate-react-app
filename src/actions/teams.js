import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_TEAM
export const addTeam = (team) => ({
  type: 'ADD_TEAM',
  team
});

export const startAddTeam = (teamData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      homeTown = '',
      teamName = '',
      note = '',
      conference = '',
      division = '',
      createdAt = 0
    } = teamData;
    const team = { homeTown, teamName, note, conference, division, createdAt };

    return database.ref(`users/${uid}/teams`).push(team).then((ref) => {
      dispatch(addTeam({
        id: ref.key,
        ...team
      }));
    });
  };
};

// REMOVE_TEAM
export const removeTeam = ({ id } = {}) => ({
  type: 'REMOVE_TEAM',
  id
});

export const startRemoveTeam = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/teams/${id}`).remove().then(() => {
      dispatch(removeTeam({ id }));
    });
  };
};

// EDIT_TEAM
export const editTeam = (id, updates) => ({
  type: 'EDIT_TEAM',
  id,
  updates
});

export const startEditTeam = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/teams/${id}`).update(updates).then(() => {
      dispatch(editTeam(id, updates));
    });
  };
};

// SET_TEAMS
export const setTeams = (teams) => ({
  type: 'SET_TEAMS',
  teams
});

export const startSetTeams = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/teams`).once('value').then((snapshot) => {
      const teams = [];

      snapshot.forEach((childSnapshot) => {
        teams.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setTeams(teams));
    });
  };
};
