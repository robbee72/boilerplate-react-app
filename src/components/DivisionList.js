import React from 'react';
import { Link } from 'react-router-dom';

const TeamListItem = ({ id, homeTown, teamName, conference, division, createdAt }) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3>{homeTown} {teamName}   {conference} {division}</h3>
  </div>
  </Link>
);

export default TeamListItem;
