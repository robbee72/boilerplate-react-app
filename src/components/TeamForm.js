import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class TeamForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      homeTown: props.team ? props.team.homeTown : '',
      teamName: props.team ? props.team.teamName : '',
      note: props.team ? props.team.note : '',
      conference: props.team ? props.team.conference : '',
      division: props.team ? props.team.division : '',
      createdAt: props.team ? moment(props.team.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onHomeTownChange = (e) => {
    const homeTown = e.target.value;
    this.setState(() => ({ homeTown }));
  };
  onTeamNameChange = (e) => {
    const teamName = e.target.value;
    this.setState(() => ({ teamName }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onConferenceChange = (e) => {
    const conference = e.target.value;

    if (!conference || conference.match) {
      this.setState(() => ({ conference }));
    }
  };
  onDivisionChange = (e) => {
    const division = e.target.value;

    if (!division || division.match) {
      this.setState(() => ({ division }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.homeTown || !this.state.teamName || !this.state.conference || !this.state.division) {
      this.setState(() => ({ error: 'Please provide Team\'s, Name, conference, and division.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        homeTown: this.state.homeTown,
        teamName: this.state.teamName,
        conference: this.state.conference,
        division: this.state.division,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Home Town"
          autoFocus
          className="text-input"
          value={this.state.homeTown}
          onChange={this.onHomeTownChange}
        />
        <input
          type="text"
          placeholder="Team Name"
          autoFocus
          className="text-input"
          value={this.state.teamName}
          onChange={this.onTeamNameChange}
        />
        <input
          type="text"
          placeholder="Conference"
          className="text-input"
          value={this.state.conference}
          onChange={this.onConferenceChange}
        />
        <input
          type="text"
          placeholder="Division"
          className="text-input"
          value={this.state.division}
          onChange={this.onDivisionChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add a note for your team (optional)"
          className="textarea"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <div>
          <button className="button">Save Team</button>
        </div>
      </form>
    )
  }
}
