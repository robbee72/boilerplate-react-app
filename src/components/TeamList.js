import React from 'react';
import { connect } from 'react-redux';
import TeamListItem from './TeamListItem';
import selectTeams from '../selectors/teams';

export const TeamList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Teams</div>
      <div className="show-for-desktop">Team</div>
      <div className="show-for-desktop">Division</div>
    </div>
    {
      props.teams.length === 0 ? (
        <p>No teams</p>
      ) : (
          props.teams.map((team) => {
            return <TeamListItem key={team.id} {...team} />;
          })
        )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    teams: selectTeams(state.teams, state.filters)
  };
};

export default connect(mapStateToProps)(TeamList);
