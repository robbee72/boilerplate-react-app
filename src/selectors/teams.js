import moment from 'moment';

// Get visible teams

export default (teams, { text, sortBy, startDate, endDate }) => {
  return teams.filter((team) => {
    const createdAtMoment = moment(team.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatchHomeTown = team.homeTown.toLowerCase().includes(text.toLowerCase());
    const textMatchTeamName = team.teamName.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatchHomeTown && textMatchTeamName;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'division') {
      return a.division < b.division ? 1 : -1;
    }
  });
};
