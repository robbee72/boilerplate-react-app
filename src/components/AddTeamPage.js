import React from 'react';
import { connect } from 'react-redux';
import TeamForm from './TeamForm';
import { startAddTeam } from '../actions/teams';

export class AddTeamPage extends React.Component {
  onSubmit = (team) => {
    this.props.startAddTeam(team);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Team</h1>
          </div>
        </div>
        <div className="content-container">
          <TeamForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddTeam: (team) => dispatch(startAddTeam(team))
});

export default connect(undefined, mapDispatchToProps)(AddTeamPage);
