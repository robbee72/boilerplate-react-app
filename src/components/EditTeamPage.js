import React from 'react';
import { connect } from 'react-redux';
import TeamForm from './TeamForm';
import { startEditTeam, startRemoveTeam } from '../actions/teams';

export class EditTeamPage extends React.Component {
  onSubmit = (team) => {
    this.props.startEditTeam(this.props.team.id, team);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemoveTeam({ id: this.props.team.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Team</h1>
          </div>
        </div>
        <div className="content-container">
          <TeamForm
            team={this.props.team}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick={this.onRemove}>Remove Team</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  team: state.teams.find((team) => team.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditTeam: (id, team) => dispatch(startEditTeam(id, team)),
  startRemoveTeam: (data) => dispatch(startRemoveTeam(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditTeamPage);
