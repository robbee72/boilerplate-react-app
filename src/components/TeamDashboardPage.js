import React from 'react';
import TeamList from './TeamList';
import TeamListFilters from './TeamListFilters';
import TeamsSummary from './TeamsSummary';

const TeamDashboardPage = () => (
  <div>
    <TeamsSummary />
    <TeamListFilters />
    <TeamList />
  </div>
);

export default TeamDashboardPage;
