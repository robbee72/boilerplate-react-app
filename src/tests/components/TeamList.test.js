import React from 'react';
import { shallow } from 'enzyme';
import { TeamList } from '../../components/TeamList';
import teams from '../fixtures/teams';

test('should render TeamList with teams', () => {
  const wrapper = shallow(<TeamList teams={teams} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render TeamList with empty message', () => {
  const wrapper = shallow(<TeamList teams={[]} />);
  expect(wrapper).toMatchSnapshot();
});
