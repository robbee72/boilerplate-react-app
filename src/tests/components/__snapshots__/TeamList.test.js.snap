// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TeamList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Teams
    </div>
    <div
      className="show-for-desktop"
    >
      Team
    </div>
    <div
      className="show-for-desktop"
    >
      Division
    </div>
  </div>
  <p>
    No teams
  </p>
</div>
`;

exports[`should render TeamList with teams 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Teams
    </div>
    <div
      className="show-for-desktop"
    >
      Team
    </div>
    <div
      className="show-for-desktop"
    >
      Division
    </div>
  </div>
  <TeamListItem
    conference="Western"
    division="Pacific"
    createdAt={0}
    homeTown="Vegas"
    id="1"
    note=""
    teamName="Golden Knights"
  />
  <TeamListItem
    conference="Western"
    division="Central"
    createdAt={-345600000}
    homeTown="Minnesota"
    id="2"
    note=""
    teamName="Wild"
  />
  <TeamListItem
    conference="Eastern"
    division="Metropolitan"
    createdAt={345600000}
    homeTown="Pittsburgh"
    id="3"
    note=""
    teamName="Penguins"
  />
</div>
`;
